File: mobile-app/src/utils/shareLinks.js
import { ensureAbsoluteUrl } from './imageUrl'

const API_ORIGIN = (() => {
  const base = import.meta.env.VITE_API_URL || 'http://localhost:5000/api'
  return base.replace(/\/api\/?$/, '')
})()

/** SPA origin (hash routes) — e.g. https://web.nexchat.site */
export function getPublicAppOrigin() {
  const configured = import.meta.env.VITE_PUBLIC_APP_URL
  if (configured && typeof configured === 'string') {
    return configured.replace(/\/$/, '')
  }
  return API_ORIGIN
}

/** API host for OG share pages (/join, /share/film, …) */
export function getShareApiOrigin() {
  return API_ORIGIN
}

export function normalizeInviteCode(code) {
  if (!code) return ''
  const c = String(code).trim().toUpperCase()
  if (!c.startsWith('NX-')) return ''
  return c.length === 7 ? c : ''
}

export function buildInvitePath(code) {
  const normalized = normalizeInviteCode(code)
  if (!normalized) return '/home'
  return `/join/${encodeURIComponent(normalized)}`
}

export function buildInviteUrl(code) {
  return buildInviteWebUrl(code)
}

export function buildInviteWebUrl(code) {
  const normalized = normalizeInviteCode(code)
  if (!normalized) return getPublicAppOrigin()
  return `${getShareApiOrigin()}/join/${encodeURIComponent(normalized)}`
}

export function buildShortFilmSharePath(filmId) {
  if (!filmId) return '/short-films'
  return `/short-films/watch?start=${encodeURIComponent(String(filmId))}`
}

export function buildShortFilmShareUrl(filmId) {
  const id = String(filmId || '').trim()
  if (!id) return `${getPublicAppOrigin()}/#/short-films`
  return `${getShareApiOrigin()}/share/film/${encodeURIComponent(id)}`
}

export function buildStorySharePath(userId) {
  if (!userId) return '/conversations'
  return `/stories/view/${encodeURIComponent(String(userId))}`
}

export function buildStoryShareUrl(userId) {
  const id = String(userId || '').trim()
  if (!id) return `${getPublicAppOrigin()}/#/conversations`
  return `${getShareApiOrigin()}/share/story/${encodeURIComponent(id)}`
}

export function buildAppHashUrl(path, query = {}) {
  const origin = getPublicAppOrigin()
  const q = new URLSearchParams()
  for (const [k, v] of Object.entries(query)) {
    if (v != null && v !== '') q.set(k, String(v))
  }
  const qs = q.toString()
  const p = path.startsWith('/') ? path : `/${path}`
  return `${origin}/#${p}${qs ? `?${qs}` : ''}`
}

/** Parse invite / share URLs (https, custom scheme, or hash routes). */
export function parseShareTargetFromUrl(rawUrl) {
  if (!rawUrl || typeof rawUrl !== 'string') return null
  let url
  try {
    url = new URL(rawUrl.replace(/^nexchat:\/\//i, 'https://nexchat.app/'))
  } catch {
    return null
  }

  const hashPath = url.hash?.startsWith('#') ? url.hash.slice(1).split('?')[0] : ''
  const path = (url.pathname || '') + (hashPath || '')

  let joinMatch = path.match(/\/join\/([A-Za-z0-9-]+)/i)
  if (!joinMatch) {
    const code = url.searchParams.get('code') || url.searchParams.get('invite')
    if (code && normalizeInviteCode(code)) {
      return { type: 'invite', code: normalizeInviteCode(code) }
    }
  } else {
    const code = normalizeInviteCode(joinMatch[1])
    if (code) return { type: 'invite', code }
  }

  const filmShareMatch = path.match(/\/share\/film\/([0-9a-f-]{36})/i)
  if (filmShareMatch?.[1]) {
    return { type: 'short_film', filmId: String(filmShareMatch[1]) }
  }
  if (path.includes('/short-films/watch')) {
    const hashQs = url.hash?.includes('?') ? url.hash.slice(url.hash.indexOf('?') + 1) : ''
    const hashParams = new URLSearchParams(hashQs)
    const id = url.searchParams.get('start') || hashParams.get('start')
    if (id) return { type: 'short_film', filmId: String(id) }
  }

  const storyMatch = path.match(/\/share\/story\/([0-9a-f-]{36})/i)
    || path.match(/\/stories\/view\/([0-9a-f-]{36})/i)
  if (storyMatch) {
    return { type: 'story', userId: String(storyMatch[1]) }
  }

  return null
}

export function absoluteMediaUrl(url) {
  if (!url) return null
  return ensureAbsoluteUrl(url)
}

File: flutter-app/lib/core/share_links.dart
import 'package:flutter/services.dart';
import 'package:share_plus/share_plus.dart';

import 'config/env.dart';
import 'i18n/i18n.dart';

/// Mirrors mobile-app/src/utils/shareLinks.js + shareExternal.js.
String normalizeInviteCode(String? code) {
  if (code == null) return '';
  final c = code.trim().toUpperCase();
  if (!c.startsWith('NX-')) return '';
  return c.length == 7 ? c : '';
}

String buildInvitePath(String? code) {
  final n = normalizeInviteCode(code);
  return n.isEmpty ? '/home' : '/join/${Uri.encodeComponent(n)}';
}

String buildInviteWebUrl(String? code) {
  final n = normalizeInviteCode(code);
  if (n.isEmpty) return Env.publicAppUrl;
  return '${Env.apiHost}/join/${Uri.encodeComponent(n)}';
}

String buildShortFilmShareUrl(String? filmId) {
  final id = (filmId ?? '').trim();
  if (id.isEmpty) return '${Env.publicAppUrl}/#/short-films';
  return '${Env.apiHost}/share/film/${Uri.encodeComponent(id)}';
}

String buildStoryShareUrl(String? userId) {
  final id = (userId ?? '').trim();
  if (id.isEmpty) return '${Env.publicAppUrl}/#/conversations';
  return '${Env.apiHost}/share/story/${Uri.encodeComponent(id)}';
}

sealed class ShareTarget {
  const ShareTarget();
}

class InviteTarget extends ShareTarget {
  const InviteTarget(this.code);
  final String code;
}

class ShortFilmTarget extends ShareTarget {
  const ShortFilmTarget(this.filmId);
  final String filmId;
}

class StoryTarget extends ShareTarget {
  const StoryTarget(this.userId);
  final String userId;
}

ShareTarget? parseShareTargetFromUrl(String? raw) {
  if (raw == null || raw.isEmpty) return null;
  Uri url;
  try {
    url = Uri.parse(raw.replaceFirst(RegExp(r'^nexchat://', caseSensitive: false), 'https://nexchat.app/'));
  } catch (_) {
    return null;
  }
  final fragment = url.fragment;
  final hashPath = fragment.split('?').first;
  final path = url.path + hashPath;
  final hashParams = fragment.contains('?') ? Uri.splitQueryString(fragment.substring(fragment.indexOf('?') + 1)) : const <String, String>{};

  final join = RegExp(r'/join/([A-Za-z0-9-]+)', caseSensitive: false).firstMatch(path);
  if (join != null) {
    final code = normalizeInviteCode(join.group(1));
    if (code.isNotEmpty) return InviteTarget(code);
  } else {
    final code = normalizeInviteCode(url.queryParameters['code'] ?? url.queryParameters['invite']);
    if (code.isNotEmpty) return InviteTarget(code);
  }

  final film = RegExp(r'/share/film/([0-9a-f-]{36})', caseSensitive: false).firstMatch(path);
  if (film != null) return ShortFilmTarget(film.group(1)!);
  if (path.contains('/short-films/watch')) {
    final id = url.queryParameters['start'] ?? hashParams['start'];
    if (id != null && id.isNotEmpty) return ShortFilmTarget(id);
  }

  final story = RegExp(r'/share/story/([0-9a-f-]{36})', caseSensitive: false).firstMatch(path) ??
      RegExp(r'/stories/view/([0-9a-f-]{36})', caseSensitive: false).firstMatch(path);
  if (story != null) return StoryTarget(story.group(1)!);
  return null;
}

/// Opens the native share sheet; falls back to the clipboard. Returns a toast message (or null).
Future<String?> shareTextPayload({String? title, required String text}) async {
  if (text.trim().isEmpty) return null;
  try {
    final r = await SharePlus.instance.share(ShareParams(text: text, subject: title, title: title));
    return r.status == ShareResultStatus.success ? t('share.shareOpened') : null;
  } catch (_) {
    await Clipboard.setData(ClipboardData(text: text));
    return t('share.linkCopied');
  }
}

Future<String?> shareInviteCode(String? code, {String? inviterName}) async {
  final n = normalizeInviteCode(code);
  if (n.isEmpty) return null;
  final link = buildInviteWebUrl(n);
  return shareTextPayload(
    title: t('share.inviteTitle'),
    text: t('share.inviteMessage', {'code': n, 'name': inviterName ?? 'NexChat', 'link': link}),
  );
}

Future<String?> shareShortFilmPublic(String filmId, {String? title}) async {
  final link = buildShortFilmShareUrl(filmId);
  final name = (title?.isNotEmpty ?? false) ? title! : t('shortFilms.title');
  return shareTextPayload(title: name, text: t('shortFilms.shareMessage', {'title': name, 'link': link}));
}

Future<String?> shareStoryPublic(String userId, {String? publisherName}) async {
  final link = buildStoryShareUrl(userId);
  final name = (publisherName?.isNotEmpty ?? false) ? publisherName! : t('stories.allStory');
  return shareTextPayload(title: t('share.storyTitle', {'name': name}), text: t('share.storyMessage', {'name': name, 'link': link}));
}
